src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,406 @@
+from arroyo.processing import StreamProcessor
+from arroyo.processing.strategies import MessageRejected
+from arroyo.processing.strategies import ProcessingStrategy
+from arroyo.processing.strategies import ProcessingStrategy as ProcessingStep
+from arroyo.processing.strategies import ProcessingStrategyFactory
+from arroyo.processing.strategies.streaming.transform import ParallelTransformStep
+from arroyo.types import Message, Partition, Position, Topic
+from django.conf import settings
+
+from sentry.utils import json, kafka_config
+
+DEFAULT_QUEUED_MAX_MESSAGE_KBYTES = 50000
+DEFAULT_QUEUED_MIN_MESSAGES = 10000
+
+logger = logging.getLogger(__name__)
+
+MessageBatch = List[Message[KafkaPayload]]
+
+
+def initializer():
+    from sentry.runner import configure
+
+    configure()
+
+
+@functools.lru_cache(maxsize=10)
+def get_indexer():
+    from sentry.sentry_metrics import indexer
+
+    return indexer
+
+
+@functools.lru_cache(maxsize=10)
+def get_task():
+    from sentry.sentry_metrics.indexer.tasks import process_indexed_metrics
+
+    return process_indexed_metrics
+
+
+@functools.lru_cache(maxsize=10)
+def get_metrics():
+    from sentry.utils import metrics
+
+    return metrics
+
+
+def get_config(topic: str, **options) -> MutableMapping[str, Any]:
+    consumer_config = kafka_config.get_kafka_consumer_cluster_options(
+        "default",
+        override_params={
+            "enable.auto.commit": False,
+            "enable.auto.offset.store": False,
+            "group.id": "ingest-metrics-consumer",
+            "default.topic.config": {"auto.offset.reset": "latest"},
+            # overridden to reduce memory usage when there's a large backlog
+            "queued.max.messages.kbytes": DEFAULT_QUEUED_MAX_MESSAGE_KBYTES,
+            "queued.min.messages": DEFAULT_QUEUED_MIN_MESSAGES,
+        },
+    )
+    return consumer_config
+
+
+class MetricsBatchBuilder:
+    """
+    Batches up individual messages - type: Message[KafkaPayload] - into a
+    list, which will later be the payload for the big outer message
+    that gets passed through to the ParallelTransformStep.
+
+    See `__flush` method of BatchMessages for when that happens.
+    """
+
+    def __init__(self, max_batch_size: int, max_batch_time: float) -> None:
+        self.__messages: MessageBatch = []
+        self.__max_batch_size = max_batch_size
+        self.__deadline = time.time() + max_batch_time
+
+    def __len__(self) -> int:
+        return len(self.__messages)
+
+    @property
+    def messages(self):
+        return self.__messages
+
+    def append(self, message: Message[KafkaPayload]) -> None:
+        self.__messages.append(message)
+
+    def ready(self) -> bool:
+        if len(self.messages) >= self.__max_batch_size:
+            return True
+        elif time.time() >= self.__deadline:
+            return True
+        else:
+            return False
+
+
+class BatchMessages(ProcessingStep[KafkaPayload]):
```

Comment:
Please unit test here as well. 
Since the pipeline is composed by discrete steps that do not really have external dependencies. They can be tested in isolation in a practical way, so I suggest you have a separate unit test per ProcessingStep you write.